apps/src/clientApi.js

```diff
@@ -68,11 +70,180 @@ class AssetsApi extends CollectionsApi {
   }
+
+  deleteFile(filename, success, error) {
+    return ajaxInternal('DELETE', this.basePath(filename), success, error);
+  }
+
+  getUploadUrl() {
+    return this.basePath('/');
+  }
+
+  makeUploadDoneCallback(callback) {
+    return callback;
+  }
+
+  makeUploadStartCallback(callback) {
+    return callback;
+  }
 }
 
+class FilesApi extends CollectionsApi {
+  getVersionHistory(success, error) {
+    var path = apiPath(
+      'files-version',
+      this.projectId || window.dashboard.project.getCurrentId()
+    );
+    return ajaxInternal('GET', path, success, error);
+  }
+
+  restorePreviousVersion(versionId, success, error) {
+    var path = apiPath(
+      'files-version',
+      this.projectId || window.dashboard.project.getCurrentId()
+    );
+    path += '?' + queryString.stringify({
+      version: versionId
+    });
+    return ajaxInternal('PUT', path, success, error);
+  }
+
+  _renameFileInternal(oldFilename, newFilename, success, error) {
+    var path = this.basePath(newFilename);
+    var params = {
+      src: oldFilename,
+      delete: oldFilename
+    };
+    if (dashboard.project.filesVersionId) {
+      params['files-version'] = dashboard.project.filesVersionId;
+    }
+    path += '?' + queryString.stringify(params);
+    return ajaxInternal('PUT', path, xhr => {
+        var response = JSON.parse(xhr.response);
+        dashboard.project.filesVersionId = response.filesVersionId;
+        if (success) {
+          success(xhr, dashboard.project.filesVersionId);
+        }
+      },
+      error);
+  }
+
+  renameFile(oldFilename, newFilename, success, error) {
+    if (this._beforeWriteHook) {
+      this._beforeWriteHook(err => {
+        // continuing regardless of error status from hook...
+        this._renameFileInternal(oldFilename, newFilename, success, error);
+      });
+      this._beforeWriteHook = null;
+    } else {
+      this._renameFileInternal(oldFilename, newFilename, success, error);
+    }
+  }
+
+  basePathWithFilesVersion(filename) {
+    var path = this.basePath(filename);
+    if (dashboard.project.filesVersionId) {
+      path += `?files-version=${dashboard.project.filesVersionId}`;
+    }
+    return path;
+  }
+
+  _deleteFileInternal(filename, success, error) {
+    return ajaxInternal('DELETE', this.basePathWithFilesVersion(filename), xhr => {
+        var response = JSON.parse(xhr.response);
+        dashboard.project.filesVersionId = response.filesVersionId;
+        if (success) {
+          success(xhr, dashboard.project.filesVersionId);
+        }
+      },
+      error);
+  }
+
+  deleteFile(filename, success, error) {
+    if (this._beforeWriteHook) {
+      this._beforeWriteHook(err => {
+        // continuing regardless of error status from hook...
+        this._deleteFileInternal(filename, success, error);
+      });
+      this._beforeWriteHook = null;
+    } else {
+      this._deleteFileInternal(filename, success, error);
+    }
```

Comment:
Good idea. I've also renamed the hook to `_beforeFirstWriteHook` to be clear that it is a one-time hook.